commit baec7c279b949a2ae207039bc92e2553b240e39d (refs/changes/80/3980/5)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-06-28T11:30:36-07:00 (1 year, 3 months ago)
    
    TRITON-52 x-DC image copy
    Reviewed by: Trent Mick <trentm@gmail.com>
    Approved by: Trent Mick <trentm@gmail.com>

diff --git a/CHANGES.md b/CHANGES.md
index 5c083cf..ddef754 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -8,6 +8,11 @@
 
 ## not yet released
 
+## 11.3.0
+
+- TRITON-52 Add imgapi.importImageFromDatacenter method, which can be used to
+  copy an image between different datacenters in the same cloud.
+
 ## 11.2.0
 
 - TRITON-53 x-account image clone. Adds the imgapi.cloneImage method.
diff --git a/lib/imgapi.js b/lib/imgapi.js
index b3bf3c9..bde4389 100644
--- a/lib/imgapi.js
+++ b/lib/imgapi.js
@@ -2579,6 +2579,95 @@ function deleteImage(uuid, account, options, callback) {
 };
 
 
+/**
+ * Imports an image from the provided datatcenter.
+ *
+ * @param {String} uuid : the UUID of the image.
+ * @param {UUID} account : The UUID of the account on behalf of whom
+ *      this request is being made.
+ * @param {Object} options: Required.
+ *      - datacenter {String} Required. The datacenter to import the image from.
+ *      - headers {Object} Optional. Additional request headers.
+ * @param {Function} callback : `function (err, job, res)`
+ */
+IMGAPI.prototype.importImageFromDatacenter =
+function importImageFromDatacenter(uuid, account, options, callback) {
+    var self = this;
+    assert.uuid(uuid, 'uuid');
+    assert.uuid(account, 'account');
+    assert.object(options, 'options');
+    assert.string(options.datacenter, 'options.datacenter');
+    assert.optionalObject(options.headers, 'options.headers');
+    assert.func(callback, 'callback');
+
+    var path = self._path(format('/images/%s', uuid), {
+        account: account,
+        action: 'import-from-datacenter',
+        channel: self.channel,
+        datacenter: options.datacenter
+    });
+    self._getAuthHeaders(function (hErr, headers) {
+        if (hErr) {
+            callback(hErr);
+            return;
+        }
+        if (options.headers) {
+            simpleMerge(headers, options.headers);
+        }
+        var reqOpts = {
+            path: path,
+            headers: headers
+        };
+        self.client.post(reqOpts, function (err, req, res, obj) {
+            if (err) {
+                callback(err, null, res);
+            } else {
+                callback(null, obj, res);
+            }
+        });
+    });
+};
+
+
+/**
+ * Imports an image from the provided datatcenter.
+ *
+ * @param {String} uuid : the UUID of the image.
+ * @param {UUID} account : The UUID of the account on behalf of whom
+ *      this request is being made.
+ * @param {Object} options: Required.
+ *      - datacenter {String} Required. The datacenter to import the image from.
+ *      - headers {Object} Optional. Additional request headers.
+ * @param {Function} callback : `function (err, image, res)`
+ */
+IMGAPI.prototype.importImageFromDatacenterAndWait =
+function importImageFromDatacenterAndWait(uuid, account, options, callback) {
+    var self = this;
+
+    self.importImageFromDatacenter(uuid, account, options,
+            function (err, job, res) {
+        if (err) {
+            callback(err, null, res);
+            return;
+        }
+
+        var wfapiUrl = res.headers['workflow-api'];
+
+        assert.string(wfapiUrl, 'wfapiUrl');
+        assert.string(job['job_uuid'], 'job_uuid');
+        assert.string(job['image_uuid'], 'image_uuid');
+
+        waitForJob(wfapiUrl, job['job_uuid'], function (jErr) {
+            if (jErr) {
+                callback(jErr);
+                return;
+            }
+            self.getImage(job['image_uuid'], callback);
+        });
+    });
+};
+
+
 /**
  * ListChannels
  * <https://mo.joyent.com/docs/imgapi/master/#ListChannels>
diff --git a/package.json b/package.json
index 0f52644..1a3b045 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-clients",
   "description": "node.js client libraries for Triton core REST APIs.",
-  "version": "11.2.0",
+  "version": "11.3.0",
   "homepage": "http://www.joyent.com",
   "repository": {
     "type": "git",
